expression/builtin_time.go

```diff
@@ -3516,10 +3516,10 @@ func (c *unixTimestampFunctionClass) getFunction(ctx sessionctx.Context, args []
 		argType := args[0].GetType()
 		argEvaltp := argType.EvalType()
 		if argEvaltp == types.ETString {
-			// Treat types.ETString as unspecified decimal.
-			retDecimal = types.UnspecifiedLength
-			if cnst, ok := args[0].(*Constant); ok {
-				tmpStr, _, err := cnst.EvalString(ctx, chunk.Row{})
+			if _, ok := args[0].(*Column); ok {
+				retDecimal = types.UnspecifiedLength
+			} else {
+				tmpStr, _, err := args[0].EvalString(ctx, chunk.Row{})
```

Comment:
Consider this case: args[0] is a scalar function, but it refers to other columns. Directly calling `args[0].EvalString()` would cause a nil pointer deference panic because the input `Row` is empty in this case.

A simple SQL can be:
```sql
select unix_timestamp(concat(a+b)) from t;
```

Reply:
You're right, I think that some scalar functions would panic. And the test passing surprises we.
Can we eval it here for scalar function like `concat`?

Reply:
For general scalar function, I think we can not. We can specially handle the `GetVar` function in the non-prepare statements. However, it's better to check how mysql handle this scenario in their codebase.

Reply:
> For general scalar function, I think we can not. We can specially handle the `GetVar` function in the non-prepare statements. However, it's better to check how mysql handle this scenario in their codebase.

Agree with you, we have better check the code in MySQL. And I think `GetVar` is not the only function we need to handle.

Reply:
A off-topic finding: `concat` seems to have bug currently
```
mysql> select concat(a+b) from t;
ERROR 1054 (42S22): Unknown column 'b' in 'field list'
mysql> select concat(a, b) from t;
ERROR 1054 (42S22): Unknown column 'b' in 'field list'
mysql> select unix_timestamp(concat(a+b)) from t;
ERROR 1054 (42S22): Unknown column 'b' in 'field list'
```
it reports error in parser.

Reply:
@eurekaka 
Do you forget the column b when create table t?
```
mysql> create table t(a int, b int);
Query OK, 0 rows affected (0.04 sec)

mysql> select concat(a+b) from t;
Empty set (0.00 sec)

mysql> select concat(a, b) from t;
Empty set (0.01 sec)

mysql> select unix_timestamp(concat(a+b)) from t;
Empty set (0.01 sec)
```

Reply:
😭 My fault.